### ISSUE TITLE
Filter changes

### ISSUE THREAD

**Issue by Keats** [Keats/tera2#37]
Here is the issue to mention:

- missing filters from v1
- wanted changes of behaviour 
- new filters wanted (no deps only)

Known missing:

- sort: do people use it? It's a lot of code

Current behaviour change:
- trim filters have been merged in trim/trim_start/trim_end with an optional `pat` argument for start/end rather than separate filters
- `int` and `float` filter do not have a default anymore
- `round` filter does not take a `common` method anymore, it's the default and should not be filled if needed
- `indent` filter now takes a `width` param rather than `prefix`
- `map`, `group_by` and `filter` filter will error if the attribute ends up being undefined on one of the value

**Comment by uncenter** [comment]
I could be wrong but it looks like you also changed `as_str` to `str`?

**Comment by uncenter** [comment]
> new filters wanted (no deps only)

- `pad` (see https://github.com/Keats/tera/pull/921)
- `repeat(n)` - repeat a string n times, result could either be an array or a joined string, though an array would probably be better for broader usecases

> wanted changes of behaviour

1. I know you have some opinions on https://github.com/Keats/tera/issues/421 though I think it should be revisited, at least concerning basic regex - no need for complicated capture groups and such.

2. It would be nice to be able to use positional args, though this isn't filters-specific as it also applies to macros. Maybe through some prefix before the argument name in the declaration, e.g. `{% macro say_hello(*name) %}` where the asterisk changes it from a kwarg to a positional arg?

3. Would it ever be possible to use https://keats.github.io/tera/docs/#map to apply a macro/filter to all elements of an array? This could be implemented without breaking existing behavior since a new kwarg could be used, e.g. `| map(func=self::say_hello)`.

**Comment by Keats** [comment]
For `repeat` I would follow what Python does, eg:

```
>>> " " * 10
'          '
>>> [0] * 10
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
>>> 
```

so the output depends on the input.

For pad, I think a `format` filter would be better so it can be used for padding but also for displaying floats with 2 decimals for example.


--- 

1. I'm not going to work on anything requiring dependencies yet but that could be added later in a `tera-contrib` crate
2. I want to keep kwargs only. It's slightly less nice when there's a single required argument but otherwise it's much more readable imo.
3. If you look at the code, there is a TODO to do exactly that. The main issue from my pov is how to pass arguments to that filter. I didn't consider macro but that could be nice as well.

**Comment by uncenter** [comment]
> For `repeat` I would follow what Python does, eg:
> 
> ```
> >>> " " * 10
> '          '
> >>> [0] * 10
> [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
> >>> 
> ```
> 
> so the output depends on the input.

Makes sense to me... are you saying you are okay adding a filter like this?

> For pad, I think a `format` filter would be better so it can be used for padding but also for displaying floats with 2 decimals for example.

Hmm how would displaying a float with 2 decimals work? I'm thinking of `pad` being used like `{{ "def" | pad(len=6, fill="_", before=true) }}` -> `"___def"`.

> 2. I want to keep kwargs only. It's slightly less nice when there's a single required argument but otherwise it's much more readable imo.

I agree but even that case where there is only a single required argument is pretty common I'd think. Besides, most programming languages *default* to positional arguments and the only language I can think of off the top of my head that does kwargs is Python, which also supports positional args.

**Comment by Keats** [comment]
> are you saying you are okay adding a filter like this?

Yes. Either that or overload `*` to do that automatically in the template without the repeat filter, one or the other.

For format, I'm thinking to basically use the Rust format: https://saghm.github.io/five-rust-things/#padding-format-operator

**Comment by azemlya** [comment]
The sort filter is very necessary. Very much so. I want to use multi-sorting, but at least like in tera v1.

**Comment by asimpletune** [comment]
Glad there's a dedicated issue for filters. I will take a closer look soon. In the meanwhile, I did notice that truncate is missing in v2. Was that intentional?

Also,  slice on strings would be convenient. Even if it can be accomplished with split | slice | join I think that's all the more reason to consider having an alias for that on strings.

**Comment by Keats** [comment]
Looks like I forgot to add truncate.

Slicing on strings is available

**Comment by Keats** [comment]
I've added the truncate filter

**Comment by asimpletune** [comment]
> I was thinking something like [array] + [array] is nicer than using a filter so probably do that instead?

O yeah, that would be nice too. 

I noticed that `indent` now is different. Instead of taking a `prefix` arg it takes just a width. I think in this case the new filter isn't as good and I prefer just being able to prefix `"     "` if I want x number of spaces. I need it because a lot of times I'm templating emails and quoting replies and so I need to prepend `> ` to every line. 

Speaking of which, a `clamp` filter would also be nice. Again with email I often want to simulate exact line widths followed by line breaks. 

I'm happy to make the changes to `filter` if you agree or add that functionality in a `clamp` filter if that also sounds like something the project would like.

**Comment by Keats** [comment]
I'm going to go with overloading `+` to concatenate arrays:

```
{{ [1, 2] + [3, 4] }}
{{ [] + [1] }}
{{ [1] + [] }}
{{ [] + [] }}
{{ [[1]] + [[2]] }}
{{ [1, 2] + [3] }}
{{ numbers + [4, 5] }}
{{ [0] + numbers }}
```
gives

```
[1, 2, 3, 4]
[1]
[1]
[]
[[1], [2]]
[1, 2, 3]
[1, 2, 3, 4, 5]
[0, 1, 2, 3]
```

**Comment by asimpletune** [comment]
Being able to use map with filters will be a godsend.

**Comment by Keats** [comment]
I think it could be added what the point of `{{ bibliography | values(key=true) }}`? Getting a tuple of (key, value) but what is the actual usecase in a template?

**Comment by jvatic** [comment]
Are there any plans to add the ability to call a function as a filter? (or is there already a way of doing this that I missed?) e.g.:

```jinja
{% set ids = [1, 2, 3, 4] %}
{% set posts = ids | map(|id| get_post(id=id)) | sort(attribute="date") %}
```

The best I've found is:

```jinja
{% set posts = [] %}
{% for id in ids %}
    {% set_global posts = [...posts, get_post(id=id)] %}
    <!-- ^ I'm not sure how to avoid using a global here -->
{% endfor %}
{% set posts = posts | sort(attribute="date") %}
```

This would simplify a bunch of template code for me. I'm happy to dig in and attempt to implement it if that'd be helpful.

**Comment by Keats** [comment]
So we have the `map` filter that allows do something like that for filters (`reviews | map(attribute="title", filter="truncate", args={"length": 5})`) but not functions yet. It's easy to add functions to it, the only question is how it looks like when you want to refer to an attribute for the arguments of the given function 🤔 
I can definitely see the need for that though

Edit: maybe having list comprehensions would be better?

**Comment by jvatic** [comment]
> So we have the map filter that allows do something like that for filters (reviews | map(attribute="title", filter="truncate", args={"length": 5})) but not functions yet. It's easy to add functions to it, the only question is how it looks like when you want to refer to an attribute for the arguments of the given function 🤔
I can definitely see the need for that though

Oh, maybe we could use `_` to reference the value map will pass in? e.g. `map(attribute="id", func="get_thing", args={"id": _, "flag": false})`. I'm not sure how that'd compose with a map filter though, unless it's easy to have `args` always be applied to a preceding `filter` or `function`? That might also get hard to follow.

And I'm not sure it belongs (exclusively) in `map` unless `attribute` becomes optional; I can see needing to call a function at any point in a filter chain, e.g. `ids | unique | <syntax to call function>`.

> Edit: maybe having list comprehensions would be better?

Do you mean something like `{% set post = [id for id in post_ids do get_post(id=id) | some_filter] | sort(attribute="date") %}` that includes piping into filters? I do like that as an option for simple cases, but I think it would get pretty difficult to read if nesting is allowed for more complex cases. And if it doesn't include filters then there'll usually be multiple assignment steps needed, which isn't terrible, but I'd prefer to get a one-liner that's easy to read.

It seems the simplest thing would be to add a filter (e.g. `call`)? Which brings us back to how to map the filter output to a function argument, and I do like using `_` for this purpose (e.g. `ids | unique | call(func="get_thing", args={"id": _, "flag": false})`). Does this seem reasonable?

Or maybe something less verbose would be better, e.g. `ids | unique |> get_thing(id=_, flag=false)` or `ids | unique | id => get_thing(id=id, flag=false)`? (I think there needs to be something to disambiguate a function from a filter, though maybe the presence of `_` is enough?). I'm not sure how much more complex this would end up being to implement, but it feels more natural.

**Comment by Keats** [comment]
> Which brings us back to how to map the filter output to a function argument, and I do like using _ for this purpose (e.g. ids | unique | call(func="get_thing", args={"id": _, "flag": false})). Does this seem reasonable?

The issue is that it doesn't scale to selecting attributes, eg `posts | map(func="get_thing", args={"id": ??})`

> Do you mean something like {% set post = [id for id in post_ids do get_post(id=id) | some_filter] | sort(attribute="date") %} that includes piping into filters? I do like that as an option for simple cases, but I think it would get pretty difficult to read if nesting is allowed for more complex cases. And if it doesn't include filters then there'll usually be multiple assignment steps needed, which isn't terrible, but I'd prefer to get a one-liner that's easy to read.

I meant something like that for your example: `{% set posts = [get_post(id=id) for id in ids] | sort(attribute="date") %}`. Similar to python. You could use any expression, so not only function or filters but even creating new objects etc ` {% set nav = [{"title": p.title, "url": p.url} for p in pages if p.in_nav] %}`. I'm 95% convinced to implement this tbh since it makes it much easier to handle patterns like the one you posted without having to cram everything into some filters.

**Comment by jvatic** [comment]
> The issue is that it doesn't scale to selecting attributes, eg posts | map(func="get_thing", args={"id": ??})

Yes, it's awkward to get something that works with that example. It'd require new syntax (e.g. wildcard ident / enum / symbol) that gets passed in to the filter.

Thinking about it a bit more and my other ideas are rather ambiguous and have implied iteration which I don't like.

> I meant something like that for your example: {% set posts = [get_post(id=id) for id in ids] | sort(attribute="date") %}. Similar to python. You could use any expression, so not only function or filters but even creating new objects etc  {% set nav = [{"title": p.title, "url": p.url} for p in pages if p.in_nav] %}. I'm 95% convinced to implement this tbh since it makes it much easier to handle patterns like the one you posted without having to cram everything into some filters.

Okay, great! Actually, I think that'd also work if I need to call a filter or two before iterating too, right? e.g. `{% set posts = [get_post(id=id) for id in (ids | unique)] %}`.

**Comment by Keats** [comment]
> I think that'd also work if I need to call a filter or two before iterating too, right? e.g. {% set posts = [get_post(id=id) for id in (ids | unique)] %}.

Yep that would work as well

**Comment by Keats** [comment]
List comprehension: https://github.com/Keats/tera2/pull/114

**Comment by Keats** [comment]
With that supported, I don't know if we still need the `map` and `filter` filters though?

**Comment by Keats** [comment]
> With that supported, I don't know if we still need the map and filter filters though?

Any thoughts on the above? I don't think they serve any purposes with list comprehensions being available so I think I would remove them before tagging v2 (hopefully next week!)

**Comment by asimpletune** [comment]
Lgtm
